File: tools/MCPackIndexer/scripts/Script_Full_Indexer.py
import os
import json
import hashlib

BASE_DIR_SCRIPT = os.path.dirname(os.path.abspath(__file__))
BASE_DIR = os.path.dirname(BASE_DIR_SCRIPT)
NEXT_DIR = os.path.join(BASE_DIR, "Pack (Next)", "assets")
MAIN_DIR = os.path.join(BASE_DIR, "Pack (Main)", "assets")
OUTPUT_DIR = os.path.join(BASE_DIR, "IndexData")

def scan_assets_folder(target_dir):
    """Scanne récursivement 100% des fichiers d'un dossier assets et retourne une liste de chemins relatifs"""
    file_list = []
    if not os.path.exists(target_dir):
        return file_list
        
    for root, dirs, files in os.walk(target_dir):
        for file in files:
            full_path = os.path.join(root, file)
            rel_path = os.path.relpath(full_path, target_dir).replace("\\", "/")
            file_list.append(rel_path)
    return sorted(file_list)

def compute_file_hash(file_path):
    """Calcule le hash MD5 d'un fichier. Pour le JSON, ignore les espaces/retour à la ligne."""
    if not os.path.exists(file_path):
        return None

    try:
        if file_path.endswith('.json') or file_path.endswith('.mcmeta'):
            with open(file_path, 'r', encoding='utf-8') as f:
                data = json.load(f)
            # Reformatage compact sans espaces et clés triées pour comparaison pure du contenu
            clean_str = json.dumps(data, separators=(',', ':'), sort_keys=True)
            return hashlib.md5(clean_str.encode('utf-8')).hexdigest()
        else:
            # Pour les autres fichiers (images png, ogg, etc.), calcul binaire direct
            hasher = hashlib.md5()
            with open(file_path, 'rb') as f:
                buf = f.read(65536)
                while len(buf) > 0:
                    hasher.update(buf)
                    buf = f.read(65536)
            return hasher.hexdigest()
    except Exception:
        # En cas d'erreur de lecture/JSON malformé, fallback en lecture binaire
        try:
            hasher = hashlib.md5()
            with open(file_path, 'rb') as f:
                buf = f.read(65536)
                while len(buf) > 0:
                    hasher.update(buf)
                    buf = f.read(65536)
            return hasher.hexdigest()
        except Exception:
            return None

def run_comparaison():
    print("\n=== DÉBUT DE L'INDEXATION COMPLÈTE & COMPARAISON ===")
    os.makedirs(OUTPUT_DIR, exist_ok=True)

    print(f"[SCAN] Analyse du dossier Vanilla : {NEXT_DIR}...")
    next_files = scan_assets_folder(NEXT_DIR)
    with open(os.path.join(OUTPUT_DIR, "Index_NextPack.json"), "w", encoding="utf-8") as f:
        json.dump(next_files, f, indent=4, ensure_ascii=False)
    print(f" -> {len(next_files)} fichiers trouvés dans Vanilla.")

    print(f"[SCAN] Analyse du dossier personnel (Pack) : {MAIN_DIR}...")
    main_files = scan_assets_folder(MAIN_DIR)
    with open(os.path.join(OUTPUT_DIR, "Index_OldPack.json"), "w", encoding="utf-8") as f:
        json.dump(main_files, f, indent=4, ensure_ascii=False)
    print(f" -> {len(main_files)} fichiers trouvés dans ton Pack.")

    next_set = set(next_files)
    main_set = set(main_files)
    
    manquants = sorted(list(next_set - main_set))
    
    comparaison_result = []
    print("[COMPARAISON] Analyse des différences de contenu via Hash MD5...")
    
    for rel_path in next_files:
        parts = rel_path.split('/')
        category = parts[1] if len(parts) > 1 else "root"
        is_missing = rel_path in manquants
        has_content_diff = False

        # Si le fichier est présent dans les deux packs, on compare les Hashs
        if not is_missing:
            path_main = os.path.join(MAIN_DIR, rel_path)
            path_next = os.path.join(NEXT_DIR, rel_path)
            
            hash_main = compute_file_hash(path_main)
            hash_next = compute_file_hash(path_next)
            
            if hash_main and hash_next and hash_main != hash_next:
                has_content_diff = True
        
        comparaison_result.append({
            "path": rel_path,
            "filename": parts[-1],
            "category": category,
            "status": "missing" if is_missing else "present",
            "has_content_diff": has_content_diff
        })

    with open(os.path.join(OUTPUT_DIR, "Index_Comparaison_Missing.json"), "w", encoding="utf-8") as f:
        json.dump(comparaison_result, f, indent=4, ensure_ascii=False)

    diff_count = sum(1 for e in comparaison_result if e.get("has_content_diff"))
    print(f" -> Comparaison terminée : {len(manquants)} fichiers manquants, {diff_count} fichiers avec du contenu différent.")
    print("=== FIN DE L'INDEXATION COMPLÈTE ===\n")

if __name__ == "__main__":
    run_comparaison()

File: tools/MCPackIndexer/Test_Conversion_JS/scripts/Script_Full_Indexer.js
const fs = require('fs');
const path = require('path');
const crypto = require('crypto');

// Définition des chemins (identique au script Python)
const BASE_DIR_SCRIPT = __dirname;
const BASE_DIR = path.dirname(BASE_DIR_SCRIPT);
const NEXT_DIR = path.join(BASE_DIR, "Pack (Next)", "assets");
const MAIN_DIR = path.join(BASE_DIR, "Pack (Main)", "assets");
const OUTPUT_DIR = path.join(BASE_DIR, "IndexData");

/**
 * Scanne récursivement 100% des fichiers d'un dossier assets
 * et retourne une liste de chemins relatifs.
 */
function scanAssetsFolder(targetDir) {
	let fileList = [];
	if (!fs.existsSync(targetDir)) {
		return fileList;
	}

	function walkSync(currentDir) {
		const entries = fs.readdirSync(currentDir, { withFileTypes: true });
		for (const entry of entries) {
			const fullPath = path.join(currentDir, entry.name);
			if (entry.isDirectory()) {
				walkSync(fullPath);
			} else if (entry.isFile()) {
				const relPath = path.relative(targetDir, fullPath).replace(/\\/g, "/");
				fileList.push(relPath);
			}
		}
	}

	walkSync(targetDir);
	return fileList.sort();
}

/**
 * Trie récursivement les clés d'un objet JS pour garantir
 * la reproductibilité de JSON.stringify (équivalent sort_keys=True)
 */
function sortObjectKeys(obj) {
	if (obj === null || typeof obj !== 'object') {
		return obj;
	}
	if (Array.isArray(obj)) {
		return obj.map(sortObjectKeys);
	}
	return Object.keys(obj)
		.sort()
		.reduce((acc, key) => {
			acc[key] = sortObjectKeys(obj[key]);
			return acc;
		}, {});
}

/**
 * Calcule le hash MD5 d'un fichier.
 * Pour le JSON/.mcmeta, reformatte en compact sans espaces avec clés triées.
 */
function computeFileHash(filePath) {
	if (!fs.existsSync(filePath)) {
		return null;
	}

	try {
		if (filePath.endsWith('.json') || filePath.endsWith('.mcmeta')) {
			const rawText = fs.readFileSync(filePath, 'utf-8');
			const data = JSON.parse(rawText);
			const sortedData = sortObjectKeys(data);
			
			// Reformatage compact sans espaces
			const cleanStr = JSON.stringify(sortedData);
			return crypto.createHash('md5').update(cleanStr, 'utf-8').digest('hex');
		} else {
			// Pour les images PNG, OGG, etc., calcul binaire direct
			const fileBuffer = fs.readFileSync(filePath);
			return crypto.createHash('md5').update(fileBuffer).digest('hex');
		}
	} catch (err) {
		// Fallback en lecture binaire directe si le JSON est malformé
		try {
			const fileBuffer = fs.readFileSync(filePath);
			return crypto.createHash('md5').update(fileBuffer).digest('hex');
		} catch (fallbackErr) {
			return null;
		}
	}
}

/**
 * Fonction principale exécutant l'indexation et la comparaison
 */
function runComparaison() {
	console.log("\n=== DÉBUT DE L'INDEXATION COMPLÈTE & COMPARAISON ===");
	fs.mkdirSync(OUTPUT_DIR, { recursive: true });

	console.log(`[SCAN] Analyse du dossier Vanilla : ${NEXT_DIR}...`);
	const nextFiles = scanAssetsFolder(NEXT_DIR);
	fs.writeFileSync(
		path.join(OUTPUT_DIR, "Index_NextPack.json"),
		JSON.stringify(nextFiles, null, 4),
		'utf-8'
	);
	console.log(` -> ${nextFiles.length} fichiers trouvés dans Vanilla.`);

	console.log(`[SCAN] Analyse du dossier personnel (Pack) : ${MAIN_DIR}...`);
	const mainFiles = scanAssetsFolder(MAIN_DIR);
	fs.writeFileSync(
		path.join(OUTPUT_DIR, "Index_OldPack.json"),
		JSON.stringify(mainFiles, null, 4),
		'utf-8'
	);
	console.log(` -> ${mainFiles.length} fichiers trouvés dans ton Pack.`);

	const nextSet = new Set(nextFiles);
	const mainSet = new Set(mainFiles);

	// Calcul des fichiers manquants dans le pack utilisateur
	const manquants = Array.from(nextSet).filter(file => !mainSet.has(file)).sort();

	const comparaisonResult = [];
	console.log("[COMPARAISON] Analyse des différences de contenu via Hash MD5...");

	for (const relPath of nextFiles) {
		const parts = relPath.split('/');
		const category = parts.length > 1 ? parts[1] : "root";
		const isMissing = manquants.includes(relPath);
		let hasContentDiff = false;

		// Si le fichier est présent dans les deux packs, on compare les MD5
		if (!isMissing) {
			const pathMain = path.join(MAIN_DIR, relPath);
			const pathNext = path.join(NEXT_DIR, relPath);

			const hashMain = computeFileHash(pathMain);
			hashNext = computeFileHash(pathNext);

			if (hashMain && hashNext && hashMain !== hashNext) {
				hasContentDiff = true;
			}
		}

		comparaisonResult.push({
			path: relPath,
			filename: parts[parts.length - 1],
			category: category,
			status: isMissing ? "missing" : "present",
			has_content_diff: hasContentDiff
		});
	}

	fs.writeFileSync(
		path.join(OUTPUT_DIR, "Index_Comparaison_Missing.json"),
		JSON.stringify(comparaisonResult, null, 4),
		'utf-8'
	);

	const diffCount = comparaisonResult.filter(e => e.has_content_diff).length;
	console.log(` -> Comparaison terminée : ${manquants.length} fichiers manquants, ${diffCount} fichiers avec du contenu différent.`);
	console.log("=== FIN DE L'INDEXATION COMPLÈTE ===\n");
}

// Exécution
runComparaison();